### ISSUE TITLE
 Error when bootstrapping CMake: build fail on LLNL linux box

### ISSUE THREAD

**Issue by jbschroder** [LLNL/spack#5806]
Ulrike and I both had the same build fail on our LLNL linux boxes.
There's apparently something wrong with our libstdc++, 

/tmp/schroder2/spack-stage/spack-stage-uSRhcV/cmake-3.9.4/Bootstrap.cmk/cmake: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /tmp/schroder2/spack-stage/spack-stage-uSRhcV/cmake-3.9.4/Bootstrap.cmk/cmake)
---------------------------------------------
Error when bootstrapping CMake:
Problem while running initial CMake
---------------------------------------------

For more details, I've attached the spack-build.out below.

Any help or comments would be much appreciated.

Thanks,

Jacob

[spack-build.out.txt](https://github.com/LLNL/spack/files/1395620/spack-build.out.txt)

**Comment by mathstuf** [comment]
CMake was built with a compiler using a new libstdc++ (looks like 4.9.2), but when running, is finding an older one (the system copy).

**Comment by ulrikeyang** [comment]
Do you know how we can help it to find a newer one?
Thanks

**Comment by mathstuf** [comment]
I wonder if the spack wrappers shouldn't also add the standard libraries in use by the wrapped compiler to the rpath as well. @tgamblin?

**Comment by tgamblin** [comment]
3 things:

1. You could try `spack install --dirty`, as this compiler looks like it requires some extra environment setup, and `--dirty` skips the initial environment cleaning Spack does to isolate builds.
2. You could try building with a spack-installed `gcc@4.9.2` by running:

    ```console
    spack install gcc@4.9.2
    spack compiler add $(spack location -i gcc@4.9.2)
    ```

    The spack compilers automatically add `RPATHs` for their runtime libraries to built executables.
3. So that this would "just work", we should look at having spack add RPATHs for runtime libraries *even for external compilers*.  This would help with `gcc` and `intel` compilers.

**Comment by mathstuf** [comment]
> So that this would "just work", we should look at having spack add RPATHs for runtime libraries even for external compilers.

Does spack skip paths which are in the global ldconfig paths already?

**Comment by tgamblin** [comment]
@mathstuf: not explicitly.  We just don't add any RPATHs for the compiler used.  We assume (maybe based on LLNL's experience) that the compilers will RAPTH their own runtime libraries.

**Comment by ulrikeyang** [comment]
it worked for me using --dirty. :-) Thanks

**Comment by tgamblin** [comment]
Ok I'll close for now but this issue came up in some other contexts (binary packaging) so I think it is worth doing (3) eventually.

**Comment by jbschroder** [comment]
"--dirty" worked for me as well.  Thanks for the help Todd!

**Comment by tgamblin** [comment]
Sorry the fix is so dirty!

**Comment by mcpherson** [comment]
Same problem 18 months later. Landed here via Google. A quick comment for others that may find themselves here for similar reasons.

I initially fixed using @tgamblin `spack install gcc@4.9.2` solution but could not leave well enough alone.

I'm using `module load gcc/6.4.0` on my system at LANL. Spack picked up the compilers from my path and put them in `~/.spack/linux/compilers.yaml`:

```yaml
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: centos7
    paths:
      cc: /projects/opt/centos7/gcc/6.4.0/bin/gcc
      cxx: /projects/opt/centos7/gcc/6.4.0/bin/g++
      f77: /projects/opt/centos7/gcc/6.4.0/bin/gfortran
      fc: /projects/opt/centos7/gcc/6.4.0/bin/gfortran
    spec: gcc@6.4.0
    target: x86_64
```

Trying `spack install cmake %gcc@6.4.0` results in the same error that @jbschroder reported:
```
/lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found
```
So, the linker is not looking for libraries in the GCC 6.4.0 path.

@rspavel advised that I modify the `compilers.yaml` file to specify the correct module:
```yaml
...
    modules:
       - gcc/6.4.0
...
```

That did the trick. ``spack install cmake %gcc@6.4.0` ran successfully.
